pkg/devfile/adapters/docker/component/adapter.go

```diff
@@ -45,7 +48,64 @@ func (a Adapter) DoesComponentExist(cmpName string) bool {
 }
 
 // Delete attempts to delete the component with the specified labels, returning an error if it fails
-// Stub function until the functionality is added as part of https://github.com/openshift/odo/issues/2581
-func (d Adapter) Delete(labels map[string]string) error {
+func (a Adapter) Delete(labels map[string]string) error {
+
+	componentName, exists := labels["component"]
+	if !exists {
+		return errors.New("unable to delete component without a component label")
+	}
+
+	list, err := a.Client.GetContainerList()
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve container list for delete operation")
+	}
+
+	componentContainer := a.Client.GetContainersByComponent(componentName, list)
+
+	if len(componentContainer) == 0 {
+		return errors.Errorf("the component %s doesn't exist", a.ComponentName)
+	}
+
+	for _, container := range componentContainer {
+		err = a.Client.RemoveContainer(container.ID)
+		if err != nil {
+			return errors.Wrapf(err, "unable to remove container ID %s of component %s", container.ID, componentName)
+		}
+
+		// Generate a list of mounted volumes of the odo-managed container
+		volumeNames := map[string]string{}
+		for _, m := range container.Mounts {
+
+			if m.Type != mount.TypeVolume {
+				continue
+			}
+			volumeNames[m.Name] = m.Name
+		}
+
+		// Locate the source volume of the container
+		volumeLabels := utils.GetProjectVolumeLabels(componentName)
+		vols, err := a.Client.GetVolumesByLabel(volumeLabels)
+		if err != nil {
+			return errors.Wrapf(err, "unable to retrieve source volume for component "+componentName)
+		}
+		if len(vols) == 0 {
+			return fmt.Errorf("unable to find source volume for component %s", componentName)
+		}
+
+		for _, vol := range vols {
+
+			// If the volume was found to be attached to the component's container, then delete the volume.
+			if _, ok := volumeNames[vol.Name]; ok {
+				err := a.Client.RemoveVolume(vol.Name)
+				if err != nil {
```

Comment:
Will this be an issue if another container is mounting the volume also, but it hasn't been stopped yet?

Would it be better to:
1) Have a list of unique volumes
2) Iterate over each container:
     a) Retrieve the container's volumes and update the list of unique volumes
     b) Remove the container
3) Remove the unique volumes found in step 2

Reply:
Agreed and updated 👍 .